# Task 7 - IGV O'clock!

Download the aligment files `pseudo_illumina_sorted.bam` , `pseudo_illumina_sorted.bam.bai` , `pseudo_illumina_sorted.bam` , `pseudo_illumina_sorted.bam.bai` and the hybrid assembly `contigs.fasta`.

Load your assembled genome - click on "Genome - Load from File..." > `contigs.fasta`

Now load your 2 alignment files - click on "Load from File..." and then select the `.bam` files.

Find a region that has decent coverage of both reads and zoom in.

![igv](https://github.com/mbtoomey/genomics_adventure/blob/release/images/chapter_5_task_7_image_1.png)

You can see that the PacBio reads are much longer, but the error rate particularly insertions and deletions is much higher than for the Illumina reads.

Explore a few other contigs to see if you can find something that looks like an error or mis-assembly. Remember the assembly process is difficult and far from perfect.

# Go to [Task 8](https://github.com/mbtoomey/genomics_adventure/blob/release/chapter_5/task_8.md)
